### PR TITLE
(proof-new) Adding rules and proof checker for EUF

### PR DESCRIPTION
**Pull request by HanielB**

### CODE REVIEW

File: src/expr/node_manager.h [hunk 1]
```diff
@@ -444,6 +444,9 @@ class NodeManager {
   /** Get a Kind from an operator expression */
   static inline Kind operatorToKind(TNode n);
 
+  /** Get corresponding application kindfor function */
```
Comment (ajreynol) [inline src/expr/node_manager.h:447]: kindfor --> kind for.
Also can you give some documentation for why this function is necessary?

File: src/expr/node_manager.h [hunk 2]
```diff
@@ -1236,6 +1239,28 @@ inline Kind NodeManager::operatorToKind(TNode n) {
   return kind::operatorToKind(n.d_nv);
 }
 
+inline Kind NodeManager::getKindForFunction(TNode fun)
```
Comment (ajreynol) [inline src/expr/node_manager.h:1242]: Move this to cpp.

File: src/expr/node_manager.h [hunk 3]
```diff
@@ -444,6 +444,9 @@ class NodeManager {
   /** Get a Kind from an operator expression */
   static inline Kind operatorToKind(TNode n);
 
+  /** Get corresponding application kindfor function */
+  static inline Kind getKindForFunction(TNode fun);
```
Comment (ajreynol) [inline src/expr/node_manager.h:448]: We don't use `inline` anymore. This is assumed to be managed by the compiler.

Comment (HanielB) [reply src/expr/node_manager.h:448]: Missed that memo. :) Will fix.

File: src/expr/node_manager.h [hunk 4]
```diff
@@ -444,7 +444,18 @@ class NodeManager {
...
-  /** Get corresponding application kindfor function */
+  /** Get corresponding application kind for function
+   *
+   * Different functional nodes are applied differently. For example,
+   * uninterpreted functions are applied via APPLY_UF, while constructors via
+   * APPLY_CONSTRUCTOR. This method provides the correct application according
+   * to which kind of function fun is.
+   *
+   * @param fun The functional node
+   * @return the correct application kind for fun. If fun is not an
```
Comment (ajreynol) [inline src/expr/node_manager.h:455]: Actually this isn't quite accurate, it just depends on the type. E.g. fun doesn't have to be an uninterpreted function, it just has to have function type?

Comment (HanielB) [reply src/expr/node_manager.h:455]: You're right. But then the more fundamental question is: who has "function type"?

Comment (ajreynol) [reply src/expr/node_manager.h:455]: well the point is that this method is based on type, not term, so just describe it that way.  You actually might even consider having this method take a TypeNode not a Node.

Comment (HanielB) [reply src/expr/node_manager.h]: I changed the doc. I think it makes sense for this to live in node manager as it refers to "how do I apply this node". Just so happens that it's solely based on its type (which whoever is calling does not necessarily need to know).

**Pull request merged**